lib/src/flowcontrol/stream_queues.dart

```diff
@@ -174,25 +174,25 @@ class StreamMessageQueueIn extends Object
     // incoming messages will get buffered.
     bufferIndicator.markBuffered();
 
-    _incomingMessagesC = new StreamController(onListen: () {
-      if (!wasClosed && !wasTerminated) {
-        _tryDispatch();
-        _tryUpdateBufferIndicator();
-      }
-    }, onPause: () {
-      _tryUpdateBufferIndicator();
-      // TODO: Would we ever want to decrease the window size in this
-      // situation?
-    }, onResume: () {
-      if (!wasClosed && !wasTerminated) {
-        _tryDispatch();
-        _tryUpdateBufferIndicator();
-      }
-    }, onCancel: () {
-      _pendingMessages.clear();
-      startClosing();
-      onCloseCheck();
-    });
+    _incomingMessagesC = new StreamController(
+        onListen: () {
+          if (!wasClosed && !wasTerminated) {
+            _tryDispatch();
+            _tryUpdateBufferIndicator();
+          }
+        },
+        onPause: () {
+          _tryUpdateBufferIndicator();
+          // TODO: Would we ever want to decrease the window size in this
+          // situation?
+        },
+        onResume: () {
+          if (!wasClosed && !wasTerminated) {
+            _tryDispatch();
+            _tryUpdateBufferIndicator();
+          }
+        },
+        onCancel: cancel);
 
     _serverPushStreamsC = new StreamController(onListen: () {
       if (!wasClosed && !wasTerminated) {
```

Comment:
Looking at the code more, it appears to me that the handling of peer pushes is not correct. We have to intercept the `cancel` event here as well and respond to `PushPromiseMessage`es with RSTs when cancelled.